src/CashuWallet.ts

```diff
@@ -967,15 +1010,71 @@ class CashuWallet {
 		secrets: Array<Uint8Array>,
 		keyset: MintKeys
 	): Array<Proof> {
-		return promises
-			.map((p: SerializedBlindedSignature, i: number) => {
-				const blindSignature = { id: p.id, amount: p.amount, C_: pointFromHex(p.C_) };
-				const r = rs[i];
-				const secret = secrets[i];
-				const A = pointFromHex(keyset.keys[p.amount]);
-				return constructProofFromPromise(blindSignature, r, secret, A);
-			})
-			.map((p: NUT11Proof) => serializeProof(p) as Proof);
+		return promises.map((p: SerializedBlindedSignature, i: number) => {
+			const dleq =
+				p.dleq == undefined
+					? undefined
+					: ({
+							s: hexToBytes(p.dleq.s),
+							e: hexToBytes(p.dleq.e),
+							r: rs[i]
+					  } as DLEQ);
+			const blindSignature = {
+				id: p.id,
+				amount: p.amount,
+				C_: pointFromHex(p.C_),
+				dleq: dleq
+			};
+			const r = rs[i];
+			const secret = secrets[i];
+			const A = pointFromHex(keyset.keys[p.amount]);
+			const proof = constructProofFromPromise(blindSignature, r, secret, A);
+			const serializedProof = serializeProof(proof) as Proof;
+			serializedProof.dleqValid =
+				dleq == undefined ? undefined : verifyDLEQProof_reblind(secret, dleq, proof.C, A);
+			serializedProof.dleq =
+				dleq == undefined
+					? undefined
+					: ({
+							s: bytesToHex(dleq.s),
+							e: bytesToHex(dleq.e),
+							r: numberToHexPadded64(dleq.r ?? BigInt(0))
+					  } as SerializedDLEQ);
+			return serializedProof;
+		});
+	}
+
+	/**
+	 * Checks that each proof in `token` has a valid DLEQ proof according to
+	 * keyset `keys`
+	 * @param token The token subject to the verification
+	 * @param keys The Mint's keyset to be used for verification
+	 */
+	private requireDLEQ(token: Token, keys: MintKeys) {
```

Comment:
I might be missing something here, but it seems like this function does not rely on any of the instance state. Therefore I would like to move it to `utils.ts` and export it to the public API. 

Functions that do not return anything, but instead throw are hard to work with IMHO. Would you not agree that it would be clearer to rename this to `hasValidDleq`, make it take a SINGLE proof, and make it return `true` is all pass and `false` if dleq are either missing or invalid? 

We could then use this util to build filters like `proofs.filter(p => hasValidDleq(p, keys)` or a util to replace `requireDleq` with a better best-case running time like 

```ts
if (proofs.some(p => !hasValidDelq(p, keys))) {
    throw new Error("Proofs contain invalid DLEQ")
}
```

Reply:
You are right.

Reply:
Doing this right now: should I let `hasValidDleq` throw in case there is no matching key from the keyset `proof.id` for a particular proof amount? Or should I just return false.

Reply:
I think it makes sense to throw in this case. `false` would indicate an invalid DLEQ, while in this case it's more than just that.

Reply:
added a test for this as well.